### PR TITLE
monitoring, Adding a new alert KubemacpoolDown

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
This PR Adds a new alert KubemacpoolDown
This alert will fire only when KMP is configured and for some reason the kubemacool-manager pod is not running.
The alert priority is set to critical, as if KMP is down, vms cannot be created/updated from the cluster.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RamLavi** [comment]
open issues:
- seems to be missing a service in order for the endpoint to be seen by the GUI
- unit tests
- e2e negative cases
- when removing CR the value stays 1. need to figure out how to zero this.

**Comment by RamLavi** [comment]
manually checked on CRC

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go [hunk 1]
```diff
@@ -286,6 +287,8 @@ func (r *ReconcileNetworkAddonsConfig) Reconcile(request reconcile.Request) (rec
 		return reconcile.Result{}, err
 	}
 
+	monitoring.TrackMonitoringComponents(&networkAddonsConfig.Spec, r.clusterInfo.MonitoringAvailable)
```
Comment (oshoval) [inline pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:290]: maybe should be after `deleteObjects` ? (actually same question is about `trackDeployedObjects` but not part of this PR)

https://github.com/kubevirt/cluster-network-addons-operator/pull/955/files#diff-463a5613afbb95588d747500e304c2d348c7fce8ec35117067322e7aac36a8dcR296

File: pkg/monitoring/metrics.go [hunk 2]
```diff
@@ -26,10 +27,15 @@ var (
 			Name: "kubevirt_cnao_cr_ready",
 			Help: "Cnao CR Ready",
 		})
+	kmpDeployedGauge = prometheus.NewGauge(
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:30]: kmp prefix is needed ? not sure

File: pkg/monitoring/metrics.go [hunk 3]
```diff
@@ -26,10 +27,15 @@ var (
...
+		prometheus.GaugeOpts{
+			Name: "kubevirt_cnao_cr_kubemacpool_deployed",
+			Help: "Kubemacpool is deployed by Cnao CR",
+		})
 )
 
 func init() {
-	metrics.Registry.MustRegister(readyGauge)
+	metrics.Registry.MustRegister(readyGauge, kmpDeployedGauge)
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:38]: maybe worth that all the gauges will be part of some slice and here we can register with `...` ?
just nice to have

File: pkg/monitoring/metrics.go [hunk 4]
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) error {
+	isKubemacpoolDeployed := false
+	if conf.KubeMacPool != nil {
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:72]: conf.KubeMacPool means its actually deployed ?
or just that it appeared on the manifest, and might be broken / just now reconcile started to deploy it ?

File: pkg/monitoring/metrics.go [hunk 5]
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
...
+	if !monitoringAvailable {
+		return nil
+	}
+
+	err := trackKubemacpoolMonitoring(conf)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:67]: return trackKubemacpoolMonitoring(conf)

**Review by oshoval — COMMENTED** [review]
First pass, thanks

Can we add some info please about the alert on the PR description ?

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:30]: what do you mean?

File: test/e2e/workflow/deployment_test.go [hunk 6]
```diff
@@ -281,25 +281,57 @@ var _ = Describe("NetworkAddonsConfig", func() {
...
-			BeforeEach(func() {
-				scrapedData, err = GetScrapedDataFromMonitoringEndpoint()
+	})
+	Context("when deploying components and checking CNAO prometheus endpoint", func() {
+		type prometheusScrapeParams struct {
+			configSpec             cnao.NetworkAddonsConfigSpec
+			expectedMetricValueMap map[string]string
+		}
+		DescribeTable("and checking scraped data",
+			func(p prometheusScrapeParams) {
```
Comment (oshoval) [inline test/e2e/workflow/deployment_test.go:291]: can the func get 2 parameters instead of the struct ?
i don't see a reason to encapsulate those parameters into a struct

File: pkg/monitoring/metrics.go (hunk 2 above)
Comment (oshoval) [reply pkg/monitoring/metrics.go:30]: i confused something, nevermind, its fine

File: pkg/monitoring/metrics.go (hunk 5 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:67]: This is intentional, I want this function to be a placeholder for future alerts.. similar to [Validate()](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/pkg/network/network.go#L27)

perhaps I should do the `errs = append(...)` as well, what do you think?

File: test/e2e/workflow/deployment_test.go (hunk 6 above)
Comment (RamLavi) [reply test/e2e/workflow/deployment_test.go:291]: I prefer it like this. 
helps keep parameter names understood , and makes thing neater imo.

File: data/monitoring/prom-rule.yaml [hunk 7]
```diff
@@ -37,3 +37,15 @@ spec:
...
             severity: warning
+        - expr: sum(up{namespace='{{ .Namespace }}', pod=~'kubemacpool-mac-controller-manager-.*'} or vector(0))
+          record: kubevirt_cnao_kubemacpool_manager_num_up_pods_total
+        - expr: sum(kubevirt_cnao_cr_kubemacpool_deployed{namespace='{{ .Namespace }}'} or vector(0))
+          record: kubevirt_cnao_cr_kubemacpool_deployed_total
+        - alert: KubemacpoolDown
+          annotations:
+            summary: Kubemacpool is deployed by CNAO CR but kubemacpool pod is down.
+            runbook_url: http://kubevirt.io/monitoring/runbooks/KubemacpoolDown
+          expr: kubevirt_cnao_cr_kubemacpool_deployed_total == 1 and kubevirt_cnao_kubemacpool_manager_num_up_pods_total == 0
```
Comment (oshoval) [inline data/monitoring/prom-rule.yaml:48]: maybe due to a bug on CNAO, or just missing kmp on the CR
kubevirt_cnao_cr_kubemacpool_deployed / kubevirt_cnao_cr_kubemacpool_deployed_total will be zero, and then the alert won't raise, it might be fine, but just raising it to be sure

File: pkg/monitoring/metrics.go [hunk 8]
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
...
+}
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) error {
+	isKubemacpoolDeployed := false
+	if conf.KubeMacPool != nil {
+		isKubemacpoolDeployed = true
+	}
+	setKubemacpoolDeployedGauge(isKubemacpoolDeployed)
+	return nil
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:76]: not needed

File: pkg/monitoring/metrics.go [hunk 9]
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
...
+
+	return nil
+}
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) error {
+	isKubemacpoolDeployed := false
+	if conf.KubeMacPool != nil {
+		isKubemacpoolDeployed = true
+	}
+	setKubemacpoolDeployedGauge(isKubemacpoolDeployed)
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:75]: if you like the style, the whole function can be one line 
setKubemacpoolDeployedGauge(conf.KubeMacPool != nil)

File: pkg/monitoring/metrics.go (hunk 5 above)
Comment (oshoval) [reply pkg/monitoring/metrics.go:67]: ye maybe then worth to do the append, else its pretty creating infra for air-conditioner :)

File: data/monitoring/prom-rule.yaml (hunk 7 above)
Comment (RamLavi) [reply data/monitoring/prom-rule.yaml:48]: the logic is that if KMP is not in the CR then KMP should *not* be deployed at all, hence the alert shouldn't fire (i.e. KMP is down but it's ok, and not something the sysAdmin should wake up in the middle of the night for).

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go (hunk 1 above)
Comment (RamLavi) [reply pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:290]: I agree it's a more fitting place.. done

File: pkg/monitoring/metrics.go (hunk 3 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:38]: I prefer it this way. it's more explicit.

File: pkg/monitoring/metrics.go (hunk 5 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:67]: Done

File: pkg/monitoring/metrics.go (hunk 9 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:75]: I prefer the variable with the long name `isKubemacpoolDeployed` to make the develper's life easier to understand..

File: pkg/monitoring/metrics.go (hunk 8 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:76]: not sure what you mean here.. the function will return []error in general just so we have the option to return errors in the future.

**Comment by RamLavi** [comment]
@oshoval [fixed](https://github.com/kubevirt/cluster-network-addons-operator/compare/ad9259124901fbb570b3ad6d0ad227269e78b04e..85a2353e3eb7a322a82e1071024c3df33c50602f) and addressed comments

File: pkg/monitoring/metrics.go (hunk 8 above)
Comment (oshoval) [reply pkg/monitoring/metrics.go:78]: i don't think we need to create infra for something that is not used right now, its a simple change that if someone need he can add a return value
atm, there is no other returns beside this one

**Comment by RamLavi** [comment]
> First pass, thanks
> 
> Can we add some info please about the alert on the PR description ?

added description in the PR

File: pkg/monitoring/metrics.go (hunk 8 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:78]: I disagree. Returning `[]error/error` from functions is generally a good practice. why not here?

Comment (oshoval) [reply pkg/monitoring/metrics.go:78]: The function has only one return statement and its always return nil
there is no good practice at all of returning the same thing no matter what the function is doing
i bet there are even static analyse tools that will warn / fail / optimize it

imo there should not be redundant code

**Comment by oshoval** [comment]
> This alert will fire only when KMP is configured and from some reason the kubemacool-manager pod is not running.

s/from/for/

**Review by oshoval — APPROVED** [review]
Thanks

File: pkg/monitoring/metrics.go (hunk 8 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:78]: I tend to prefer adding the infra beforehand.
`TrackMonitoringComponents()` is the house for handling monitoring for all deployed components.
Each can potentially return an error, so `TrackMonitoringComponents()` collects the errors from them and flattens them to one error line if needed. 
Granted, currently there is only one: `trackKubemacpoolMonitoring()`, but I prefer that we lay out how we want to function to look like from the beginning, so that next time we want to add another monitoring (for knmstate for example) then that developer would only need to add the component's function.

We did the same when we added the `SpecialCleanUp()` [here](https://github.com/kubevirt/cluster-network-addons-operator/pull/270), and I think it's what we should do here as well.

**Comment by RamLavi** [comment]
Added a function `ResetMonitoringComponents()` to reset the gouges when the CR is removed [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/85a2353e3eb7a322a82e1071024c3df33c50602f..3516a4c9151adc5698758b7bfefa2c319483dad7)

File: pkg/monitoring/metrics.go [hunk 10]
```diff
@@ -40,6 +47,42 @@ func SetReadyGauge(isReady bool) {
...
+		kmpDeployedGauge.Set(1)
+	} else {
+		kmpDeployedGauge.Set(0)
+	}
+}
+
+func ResetMonitoringComponents() error {
+	SetReadyGauge(false)
+	setKubemacpoolDeployedGauge(false)
+	return nil
```
Comment (oshoval) [inline pkg/monitoring/metrics.go:61]: same deal here, i think return nil for a function that doesn't do anything else is totally unneeded,
its ok that a function won't return a value

maybe worth that the SetReadyGauge will be on other PR / commit, but i won't insist

**Review by oshoval — APPROVED** [review]

File: pkg/monitoring/metrics.go (hunk 10 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:61]: in this case I tend to agree there is no need for error in the function. will remove

**Review by oshoval — APPROVED** [review]

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go [hunk 11]
```diff
@@ -297,6 +300,8 @@ func (r *ReconcileNetworkAddonsConfig) Reconcile(request reconcile.Request) (rec
...
```
Comment (phoracek) [inline pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:303]: The condition MonitoringAvailable should be checked here. If it is not available, we should not invoke the module at all.  That way we encapsulate the helper module and isolate it from the dependencies. Would that make sense?

File: data/monitoring/prom-rule.yaml [hunk 12]
```diff
@@ -37,3 +37,15 @@ spec:
           for: 5m
           labels:
             severity: warning
+        - expr: sum(up{namespace='{{ .Namespace }}', pod=~'kubemacpool-mac-controller-manager-.*'} or vector(0))
+          record: kubevirt_cnao_kubemacpool_manager_num_up_pods_total
+        - expr: sum(kubevirt_cnao_cr_kubemacpool_deployed{namespace='{{ .Namespace }}'} or vector(0))
```
Comment (phoracek) [inline data/monitoring/prom-rule.yaml:42]: Why do we need the sum here? Can't we just remove true or false from kubevirt_cnao_cr_kubemacpool_deployed ? We do not care about the number of deployed objects here, do we? Only about whether it is enabled or not

File: pkg/monitoring/metrics.go [hunk 13]
```diff
@@ -40,6 +47,41 @@ func SetReadyGauge(isReady bool) {
...
 
+func setKubemacpoolDeployedGauge(isDeployed bool) {
+	if isDeployed {
+		kmpDeployedGauge.Set(1)
+	} else {
+		kmpDeployedGauge.Set(0)
+	}
+}
+
+func ResetMonitoringComponents() {
```
Comment (phoracek) [inline pkg/monitoring/metrics.go:58]: Are not these monitored components?

File: pkg/monitoring/metrics.go [hunk 14]
```diff
@@ -40,6 +47,41 @@ func SetReadyGauge(isReady bool) {
...
+	errs = append(errs, trackKubemacpoolMonitoring(conf)...)
+
+	if len(errs) > 0 {
+		return errors.Errorf("Error in tracking monitoring components:\n%s", errorListToMultiLineString(errs))
+	}
+	return nil
+}
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) []error {
+	isKubemacpoolDeployed := false
```
Comment (phoracek) [inline pkg/monitoring/metrics.go:77]: Would this work?

```go
isKubemacpoolDeployed := conf.KubeMacPool != nil && conf.KubeMacPool
```

File: pkg/monitoring/metrics.go (hunk 9 above)
Comment (phoracek) [reply pkg/monitoring/metrics.go:81]: I am concerned about this hidden side effect. Could we either somehow lift this to an upper-level function or to rename this function to clearly show that it is changing the global? Maybe just use the setKubemacpoolDeployedGauge name directly? And if you could pass reference to the global to the function, it would get rid of the side effect completely

**Review by phoracek — CHANGES_REQUESTED** [review]

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go (hunk 11 above)
Comment (RamLavi) [reply pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:303]: I don't mind, but note we do it differently on render functions (see [here](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/pkg/network/kubemacpool.go#L62-L64))
Do you think we should change it there as well? (in a different PR of course)

File: data/monitoring/prom-rule.yaml (hunk 12 above)
Comment (RamLavi) [reply data/monitoring/prom-rule.yaml:42]: the alert is using these 2 records `if(x==1 && y==0) for >5min then fire!` 
In order for the alert to work, the records needs to point to the same entry (x counts apples (arbitrary gauge) and y counts bananas (num of up kmp pods) )
After some playing with this with @sradco, we saw that `sum()` normalizes the records to be of same in "type" and so it works.

File: pkg/monitoring/metrics.go (hunk 13 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:58]: They are! Thank you!!! 
finding the right wording brings me joy =)

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go (hunk 11 above)
Comment (RamLavi) [reply pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:303]: Done

File: pkg/monitoring/metrics.go (hunk 13 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:58]: Done

File: pkg/monitoring/metrics.go (hunk 14 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:77]: like! Done

File: pkg/monitoring/metrics.go (hunk 9 above)
Comment (RamLavi) [reply pkg/monitoring/metrics.go:81]: I see, I'll move it upward, but not sure why I should pass the global. can you elaborate?

Comment (phoracek) [reply pkg/monitoring/metrics.go:81]: If you pass link to the global, you can unit test it. It would also become clear what is being "set". With it being taken from the global context, we cannot easily unit test it. It is unclear what this function works with unless you peek into it.

Comment (phoracek) [reply pkg/monitoring/metrics.go:64]: Sounds good

**Review by phoracek — APPROVED** [review]
/lgtm
/approve
